### PR TITLE
enable UseDataClass rule for detekt code base

### PR DESCRIPTION
**Pull request by marschwar**
Belongs to discussion #3993

### CODE REVIEW

**Review by cortinico — APPROVED** [review]

File: detekt-metrics/src/main/kotlin/io/github/detekt/metrics/ComplexityMetric.kt [hunk 1]
```diff
@@ -6,14 +6,30 @@ import io.github.detekt.metrics.processors.linesKey
...
 
-    val mcc = detektion.getData(complexityKey)
-    val cognitiveComplexity = detektion.getData(CognitiveComplexity.KEY)
-    val loc = detektion.getData(linesKey)
-    val sloc = detektion.getData(sourceLinesKey)
-    val lloc = detektion.getData(logicalLinesKey)
-    val cloc = detektion.getData(commentLinesKey)
-    val findings = detektion.findings.entries
+    companion object Factory {
+        fun create(detektion: Detektion): ComplexityMetric = detektion.run {
```
Comment (BraisGabin) [inline detekt-metrics/src/main/kotlin/io/github/detekt/metrics/ComplexityMetric.kt:23]: I never know if in this case is better to use `run` or `with`. Both does the same but I don't know which one is more semantically correct. 

And, another think is factory vs secundary constructor.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/Lifecycle.kt [hunk 2]
```diff
@@ -54,7 +54,7 @@ internal interface Lifecycle {
     }
 }
 
-internal class DefaultLifecycle(
+internal data class DefaultLifecycle(
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/Lifecycle.kt:57]: This class doesn't feel like a data class, don't you think? I don't know if it's a false positive or just a good place to add a suppress.

If this is a false positive how could we fix it? Maybe we could check if all the properties implement a real equals? Or that the properties are data classes?

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektReport.kt [hunk 3]
```diff
@@ -2,7 +2,7 @@ package io.gitlab.arturbosch.detekt.extensions
 
 import java.io.File
 
-class DetektReport(val type: DetektReportType) {
+data class DetektReport(val type: DetektReportType) {
```
Comment (BraisGabin) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektReport.kt:5]: And I think that this is a real false positive. This class has other properties so the equals can't be correctly implemented.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/Lifecycle.kt (hunk 2 above)
Comment (marschwar) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/Lifecycle.kt:57]: I am actually not worried so much about the equals method (maybe I should) but rather what the class does. If it just holds data, it should be a data class. At first glance this seems to be the case. But the class implements an interface with a function and therefore does have behavior. Hence you are right - it should not be a data class. But it really is hard to define a rule for this.

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektReport.kt (hunk 3 above)
Comment (marschwar) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektReport.kt:5]: Good catch. But I don't think it is a false positive. I added those properties to the primary constructor. All this class does is hold data.

File: detekt-metrics/src/main/kotlin/io/github/detekt/metrics/ComplexityMetric.kt [hunk 4]
```diff
@@ -6,14 +6,30 @@ import io.github.detekt.metrics.processors.linesKey
 import io.github.detekt.metrics.processors.logicalLinesKey
 import io.github.detekt.metrics.processors.sourceLinesKey
 import io.gitlab.arturbosch.detekt.api.Detektion
+import io.gitlab.arturbosch.detekt.api.Finding
+import io.gitlab.arturbosch.detekt.api.RuleSetId
 
-class ComplexityMetric(detektion: Detektion) {
+data class ComplexityMetric(
```
Comment (schalkms) [inline detekt-metrics/src/main/kotlin/io/github/detekt/metrics/ComplexityMetric.kt:12]: I'm not sure whether the new code is better overall than the old one.

**Comment by cortinico** [comment]
> So my vote goes to keep just the changes related with those two files and don't enable this rule by default in the detekt code.

Agree. Specifically the `ComplexityMetric` change seems like is not improving readability.

**Pull request closed without merging**